services/repository/transfer.go

```diff
@@ -42,12 +42,17 @@ func TransferOwnership(ctx context.Context, doer, newOwner *user_model.User, rep
 
 	oldOwner := repo.Owner
 
-	repoWorkingPool.CheckIn(fmt.Sprint(repo.ID))
+	ctx, releaser, err := globallock.Lock(ctx, getRepoWorkingLockKey(repo.ID))
+	if err != nil {
+		log.Error("lock.Lock(): %v", err)
+		return fmt.Errorf("lock.Lock: %w", err)
+	}
+
 	if err := transferOwnership(ctx, doer, newOwner.Name, repo); err != nil {
-		repoWorkingPool.CheckOut(fmt.Sprint(repo.ID))
+		releaser()
 		return err
 	}
-	repoWorkingPool.CheckOut(fmt.Sprint(repo.ID))
+	releaser()
```

Comment:
It's not right. When you release the lock, the ctx returned by `globallock.Lock` will be canceled. See the example of "Early release": https://github.com/go-gitea/gitea/pull/31908

It should be:

```go
	ctx, releaser, err := globallock.Lock(ctx, getRepoWorkingLockKey(repo.ID))
	if err != nil {
		log.Error("lock.Lock(): %v", err)
		return fmt.Errorf("lock.Lock: %w", err)
	}
	defer releaser()

	if err := transferOwnership(ctx, doer, newOwner.Name, repo); err != nil {
		return err
	}
	ctx = releaser()
```

(Why don't apply suggestions:
<img width="322" alt="image" src="https://github.com/user-attachments/assets/d95203c6-7dd8-44f3-8675-8bb73c79a54b">
)

Reply:
I think that's why the CI failed
<img width="1484" alt="image" src="https://github.com/user-attachments/assets/050a42c7-afa9-41b7-90dc-a2e092b05b15">

Reply:
> It's not right. When you release the lock, the ctx returned by `globallock.Lock` will be canceled. See the example of "Early release": #31908

I would say it is an over-design.

I can't see any real use-case for using such "ctx with releaser", and it will be frequently abused in the future.

Reply:
I mean `Lock` shouldn't return a new context which could be canceled by "release()".

Just make `func Lock() (releaser, error)`

Reply:
I see. It's a way let the caller stop when they have lost the lock. See #31908 "Acquired locks could be lost"

Reply:
If "Acquired locks could be lost", a lot of things will wrong, and I do not think it's able or worth to handle such case.

For example: even if some code really uses the lock ctx, many operations are not revert-able, if the lock gets lost, the code still goes wrong.

Distro-lock doesn't really fit complex business code, unless every line of code is carefully designed to be distro-safe.

Reply:
Let's see a real example (take the `transferOwnership` in your example):

`transferOwnership` does `util.Rename(oldRepoPath, newRepoPath)`, if you cancel the lock ctx, then the database operations are rolled-back, but the paths on the storage have been renamed, them, Gitea gets corrupted.

Reply:
I understand, and I also agree distro-lock doesn't really fit complex business code.

And my viewpoint is that if an operation cannot work well with canceled context, it shouldn't be the lock's fault.

A function defined with a context as parameter cannot declare that "never cancel the input ctx before it return". There could be many things causing the context to be canceled, like the deadline has arrived, the request has been canceled, or even the process has exited.

If the function will lead to a very bad situation if it has been canceled, it should find a way to improve that, like don't accept input ctx.

---

"Acquired locks could be lost", it's not only the design of the module, but also the objective fact of distro-lock. The module just try no let that callers know that it happened. If the callers believe that things should go on even it lost the lock, the code could be:

```go
	// Don't overwrite the ctx since we don't want to stop transferOwnership even if it lost the lock
	_, releaser, err := globallock.Lock(ctx, getRepoWorkingLockKey(repo.ID))
	if err != nil {
		log.Error("lock.Lock(): %v", err)
		return fmt.Errorf("lock.Lock: %w", err)
	}
	defer releaser()

	if err := transferOwnership(ctx, doer, newOwner.Name, repo); err != nil {
		return err
	}
	// ...
```

Reply:
> If the callers believe that things should go on even it lost the lock, the code can be:

In real world, I do not see a real case that "callers must respect the lost lock". That's why I think the "ctx with releaser" is an over-design and doesn't really help.

Reply:
I see, thanks for bringing me to think about it more deeply. And I believe the current solution may not be perfect, but at least acceptable.